gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/strategy/optimization/FilterRankingStrategy.java

```diff
@@ -232,4 +243,64 @@ public Set<Class<? extends OptimizationStrategy>> applyPrior() {
     public static FilterRankingStrategy instance() {
         return INSTANCE;
     }
+
+    /**
+     * Get steps of the specified class throughout the traversal and grouping them based on the traversal parent
+     * collecting them in a fashion that orders them from the deepest steps first
+     */
+    public static List<Pair<TraversalParent, List<Step<?,?>>>> collectStepsOfAssignableClassRecursivelyFromDepthGroupedByParent(
+            final Traversal.Admin<?, ?> traversal, final Class<?> stepClass) {
+
+        final List<Pair<TraversalParent, List<Step<?,?>>>> collectionList = new ArrayList<>();
+        final Stack<Step<?,?>> stack = new Stack<>();
```

Comment:
nit: not informative variable names. may be something like `unprocessedSteps` will be better?